### ISSUE TITLE
Make is clearer when the program is loading

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#1390]
Right now it appears as if some compiled programs run very slowly, but much of that waiting time is spent downloading the compiled code.  To give the end-user a better feel for the behavior of their own code, it would be better if the environment made it clear when the code was done loading and actually started.

Fortunately, this isn't too difficult!  The compiled code already sends a `programStarted` message via `window.postMessage` when the code is fully loaded and starts to run.  Right now, that's used only for a timer that decides how to report deferred errors.  We could, instead, continue to show the compile wait dialog and dismiss it only when `programStarted` is received.  Would also have to be careful to always dismiss when something goes wrong.

**Comment by cdsmith** [comment]
This isn't a big deal most of the time.  But, for example, I was doing a demo of different ways to compute prime numbers, and the whole point is that some of them are faster than others.  In GHCi, it's clear when one runs in twice the time of the other.  But in CodeWorld, it's lost in the noise of loading the compiled code.  That's frustrating.

**Comment by cdsmith** [comment]
I played with this for a bit.  It looks like a good part of the startup time isn't loading.  It's some kind of delay inside GHCJS runtime, between the call to `window.h$run(window.h$mainZCZCMainzimain);` and the first line of `main` running in the student code.  That's too bad.

I grabbed a profile.  There's only 75 ms of actual script running, which is mostly h$initInfoTables.  But then there's about 2 seconds during which no scripts are running at all, before `main` starts to execute.  It's not clear what's going on during that time period.  It looks like when the program finally starts, it's because a timer fires.  That timer is set in the GHCJS runtime here:

```
function h$scheduleMainLoop() {
    ;
    if(h$mainLoopImmediate) return;
    h$clearScheduleMainLoop();
    if(h$delayed.size() === 0) {
            ;
            h$mainLoopTimeout = setTimeout(h$mainLoop, h$gcInterval);
        return;
    }
```

I'm way out of my depth here.  (@luite , any help?)  But it appears we're waiting on a `h$gcInterval` (actually, maybe two of them) before running `main`.

**Comment by cdsmith** [comment]
I guess one hacky way around this would be to have run.js ramp down `h$gcInterval` to something tiny, and then do a `setTimeout` to increase it again once the program is in progress?  That seems like the wrong answer, though.

**Comment by cdsmith** [comment]
I tried the work-around, setting `h$gcInterval` to 100 before calling `h$run`, and it seems to work.  Here's what I have now:

* 80 ms in `h$runInitStatic`
* A delay of exactly 100 ms (previously, this was 1000 ms).
* 1.2 ms for the GC itself.
* Another delay of exactly 100 ms (previously, this was 1000 ms).
* The program starts about exactly 200 ms after `h$runInitStatic` finishes.

I successfully set `h$gcInterval` to 25, too, which reduces the delay to 50 ms.  Since the initial `h$mainLoop` takes 80 ms on its own, there's no real benefit to reducing it any further with this hack.  Being too aggressive could result in GC thrashing on low-powered clients.  Further work here should focus on:

1. Fixing the underlying issue with the GHCJS runtime, if it's not too hard.
2. Making the UI change that I originally proposed here, which delays dismissed the "Compiling..." alert window until the `notifyStarted` message is received.

**Comment by Powell-v2** [comment]
I believe there is more to this task than just closing an alert when the [`programStarted` message](https://github.com/google/codeworld/blob/master/web/js/codeworld.js#L950) is received?

**Comment by cdsmith** [comment]
That's the essence of it.  The alert currently closes when the `run.html` URL is set on the iframe.  Intead, it should remain open when the program is started, and it should close when the `programStarted` message is received.  One corner case to handle is what to do if the user clicks Cancel between the times that the DOM is changed, and the `programStarted` message is received.  I think Cancel should act like Stop in that case.

**Comment by Powell-v2** [comment]
Okay, got it.

By the way, I was trying out various examples and one of them, the Mandelbrot Set, almost brings page to a halt. Is that normal?

**Comment by cdsmith** [comment]
Ah, the mandelbrot example has recursion set too high.  I'll fix that.
